book/src/internals/type-system.md

```diff
@@ -0,0 +1,323 @@
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Theory
+
+> For any undefined terms used in this section, refer to set theory and
+> mathematical definitions in general.
+
+A "type of a variable" is a "set of all possible values of that variable". This
+means that terms "type" and "set" are equivalent in this context.
+
+Types (sets) can be expressions. For example, a union of two types is a type
+itself. This means a type expression is equivalent to any other expression whose
+type is a "set of sets".
+
+So let's introduce a "set" as a PRQL expression construct (alongside existing
+idents, literals, ranges and so on). For now, it does not need any special
+syntax. Because sets are normal expressions, existing syntax can be repurposed
+to define operations on sets:
+
+- Binary operation `or` of two sets represents a union of those two sets:
+
+  ```
+  let number = int or float
+  ```
+
+  With algebraic types, this is named "a sum type".
+
+- Literals can be coerced into a singleton set (i.e. `false` is converted into a
+  set with only one element `false`):
+
+  ```
+  let int_or_null = int or null
+  ```
+
+- A list of set expressions can be coerced into a set of tuples, where entries
+  of the tuples correspond to elements of the set expressions in the list:
+
+  ```
+  let my_row = [id = int, bool, name = str]
+  ```
+
+- An array of set expressions with exactly one entry can be coerced into a set
+  of arrays of that set expression:
+  ```
+  let array_of_int = {int} # proposed syntax for arrays
+  ```
+
+Module `std` defines built-in sets `int`, `float`, `bool`, `text` and `set`.
+Other built-in sets will be added in the future.
+
+## Type annotations
+
+Let's extend the syntax for declaration of variable `a`, whose value can be
+computed by evaluating `x`, with a type annotation:
+
+```
+let a <t> = x
+```
+
+This extended syntax applies following assertions:
+
+- `t` can be evaluated statically (at compile time),
+- `t` can be coerced into a set,
+- value of `x` (and `a`) must be an element of `t`. This assertion must be
+  possible to evaluate statically.
+
+Similar rules apply to type annotations of return types of functions and
+function parameter definitions.
+
+## Type definitions
+
+As shown, types can be defined by defining expressions and coercing them to set
+expressions by using `< >`.
+
+But similar to how both `func` and `let` can be used to define functions (when
+we introduce lambda function syntax), let's also define syntactic sugar for type
+definitions:
+
+```
+# these two are equivalent
+let my_type <set> = set_expr
+type my_type = set_expr
+```
```

Comment:
FWIW I'm increasingly in favor of this approach — i.e. functions can be `let add = a b -> a + b`

OTOH having two ways of defining these is not ideal. So we could only do the first, and lint towards having these for set/types & functions...